### ISSUE TITLE
Backend not found during initialization

### ISSUE THREAD

**Issue by hadim** [deeplearning4j/nd4j#2248]
I am building an ImageJ plugin that uses your library. Everything is fine when I start ImageJ from Eclipse,

Now, when I run `mvn clean install`, the plugin and all its dependencies are copied to the ImageJ directory. If I start ImageJ and start the plugin I have the following error:

```
Caused by: java.lang.ExceptionInInitializerError
    at fiji.plugin.filamentdetector.analyzer.NaiveNucleationAnalyzer.analyze(NaiveNucleationAnalyzer.java:96)
    at fiji.plugin.filamentdetector.gui.controller.AnalyzeController.analyze(AnalyzeController.java:222)
    ... 69 more
Caused by: java.lang.RuntimeException: org.nd4j.linalg.factory.Nd4jBackend$NoAvailableBackendException: Please ensure that you have an nd4j backend on your classpath. Please see: http://nd4j.org/getstarted.html
    at org.nd4j.linalg.factory.Nd4j.initContext(Nd4j.java:6089)
    at org.nd4j.linalg.factory.Nd4j.<clinit>(Nd4j.java:201)
    ... 71 more
Caused by: org.nd4j.linalg.factory.Nd4jBackend$NoAvailableBackendException: Please ensure that you have an nd4j backend on your classpath. Please see: http://nd4j.org/getstarted.html
    at org.nd4j.linalg.factory.Nd4jBackend.load(Nd4jBackend.java:258)
    at org.nd4j.linalg.factory.Nd4j.initContext(Nd4j.java:6086)
    ... 72 more
```

I checked and all the related ND4J JAR files are correctly copied to the ImageJ installation.

I don't really know if it's an ND4J bug or an ImageJ one or if I am just doing something wrong... So let me know if you have an idea.

Also pinging @ctrueden (lead developer of ImageJ).

Related issue: https://github.com/hadim/FilamentDetector/issues/8

**Comment by AlexDBlack** [comment]
Is it possible to get a dependency tree or something here?
Also, I assume you are including a backend - nd4j-native-platform, for example, that is present on the classpath when running your program?

**Comment by hadim** [comment]
Yes the nd4j-native-platform are all copied to the imageJ installation:

```
ls -l nd4j-native-*
-rw-r--r-- 1 hadim hadim 2142555 Nov  6 16:20 nd4j-native-0.9.1-android-arm.jar
-rw-r--r-- 1 hadim hadim 2639726 Nov  6 16:20 nd4j-native-0.9.1-android-x86.jar
-rw-r--r-- 1 hadim hadim  791820 Nov  7 16:49 nd4j-native-0.9.1-linux-ppc64le.jar
-rw-r--r-- 1 hadim hadim  963237 Nov  6 16:20 nd4j-native-0.9.1-linux-x86_64.jar
-rw-r--r-- 1 hadim hadim 1434741 Nov  6 16:20 nd4j-native-0.9.1-macosx-x86_64.jar
-rw-r--r-- 1 hadim hadim 2127226 Nov  6 16:20 nd4j-native-0.9.1-windows-x86_64.jar
```

Here is a link my pom.xml: https://github.com/hadim/FilamentDetector/blob/master/pom.xml#L95

Here is a dependency tree:

```
$ mvn dependency:tree
[INFO] Scanning for projects...
[INFO] 
[INFO] ------------------------------------------------------------------------
[INFO] Building FilamentDetector 0.3.4-SNAPSHOT
[INFO] ------------------------------------------------------------------------
[INFO] 
[INFO] --- maven-dependency-plugin:3.0.0:tree (default-cli) @ FilamentDetector ---
[INFO] org.scijava:FilamentDetector:jar:0.3.4-SNAPSHOT
[INFO] +- net.imagej:imagej:jar:2.0.0-rc-61:compile
[INFO] |  +- net.imagej:imagej-common:jar:0.24.4:compile
[INFO] |  |  \- edu.ucar:udunits:jar:4.3.18:compile
[INFO] |  +- net.imagej:imagej-notebook:jar:0.2.2:compile
[INFO] |  +- net.imagej:imagej-ops:jar:0.38.0:compile
[INFO] |  |  +- net.imglib2:imglib2-algorithm-fft:jar:0.1.4:compile
[INFO] |  |  +- org.scijava:parsington:jar:1.0.1:compile
[INFO] |  |  \- org.scijava:scripting-javascript:jar:0.4.4:compile
[INFO] |  +- net.imagej:imagej-updater:jar:0.8.3:compile
[INFO] |  +- io.scif:scifio:jar:0.32.0:compile
[INFO] |  |  +- io.scif:scifio-jai-imageio:jar:1.1.1:compile
[INFO] |  |  \- org.mapdb:mapdb:jar:1.0.3:compile
[INFO] |  +- org.scijava:scijava-common:jar:2.65.0:compile
[INFO] |  |  +- com.googlecode.gentyref:gentyref:jar:1.1.0:compile
[INFO] |  |  \- org.bushe:eventbus:jar:1.4:compile
[INFO] |  +- net.imagej:imagej-deprecated:jar:0.1.2:compile
[INFO] |  +- net.imagej:imagej-plugins-commands:jar:0.7.1:runtime
[INFO] |  |  +- com.github.sbridges.object-inspector:object-inspector:jar:0.1:runtime
[INFO] |  |  \- net.iharder:base64:jar:2.3.8:runtime
[INFO] |  +- net.imagej:imagej-plugins-tools:jar:0.3.1:runtime
[INFO] |  +- net.imagej:imagej-plugins-uploader-ssh:jar:0.3.2:runtime
[INFO] |  |  \- com.jcraft:jsch:jar:0.1.54:runtime
[INFO] |  +- net.imagej:imagej-plugins-uploader-webdav:jar:0.2.2:runtime
[INFO] |  +- net.imagej:imagej-scripting:jar:0.6.0:runtime
[INFO] |  |  +- org.scijava:scripting-beanshell:jar:0.3.2:runtime
[INFO] |  |  |  \- org.apache-extras.beanshell:bsh:jar:2.0b6:runtime
[INFO] |  |  +- org.scijava:scripting-clojure:jar:0.1.6:runtime
[INFO] |  |  |  \- org.clojure:clojure:jar:1.8.0:runtime
[INFO] |  |  +- org.scijava:scripting-groovy:jar:0.2.6:runtime
[INFO] |  |  |  +- org.codehaus.groovy:groovy:jar:2.4.8:runtime
[INFO] |  |  |  \- org.apache.ivy:ivy:jar:2.2.0:runtime
[INFO] |  |  +- org.scijava:scripting-java:jar:0.4.1:runtime
[INFO] |  |  |  \- org.scijava:minimaven:jar:2.2.1:runtime
[INFO] |  |  +- org.scijava:scripting-jruby:jar:0.2.5:runtime
[INFO] |  |  |  +- org.jruby:jruby-core:jar:1.7.12:runtime
[INFO] |  |  |  |  +- org.ow2.asm:asm:jar:4.0:runtime
[INFO] |  |  |  |  +- org.ow2.asm:asm-commons:jar:4.0:runtime
[INFO] |  |  |  |  |  \- org.ow2.asm:asm-tree:jar:4.0:runtime
[INFO] |  |  |  |  +- org.ow2.asm:asm-analysis:jar:4.0:runtime
[INFO] |  |  |  |  +- org.ow2.asm:asm-util:jar:4.0:runtime
[INFO] |  |  |  |  +- org.jruby.joni:joni:jar:2.1.1:runtime
[INFO] |  |  |  |  +- com.github.jnr:jnr-netdb:jar:1.1.2:runtime
[INFO] |  |  |  |  +- com.github.jnr:jnr-enxio:jar:0.4:runtime
[INFO] |  |  |  |  +- com.github.jnr:jnr-x86asm:jar:1.0.2:runtime
[INFO] |  |  |  |  +- com.github.jnr:jnr-unixsocket:jar:0.3:runtime
[INFO] |  |  |  |  +- com.github.jnr:jnr-posix:jar:3.0.1:runtime
[INFO] |  |  |  |  +- org.jruby.extras:bytelist:jar:1.0.11:runtime
[INFO] |  |  |  |  +- com.github.jnr:jnr-constants:jar:0.8.5:runtime
[INFO] |  |  |  |  +- org.jruby.jcodings:jcodings:jar:1.0.10:runtime
[INFO] |  |  |  |  +- com.github.jnr:jnr-ffi:jar:1.0.7:runtime
[INFO] |  |  |  |  +- com.github.jnr:jffi:jar:1.2.7:runtime
[INFO] |  |  |  |  +- com.github.jnr:jffi:jar:native:1.2.7:runtime
[INFO] |  |  |  |  +- com.jcraft:jzlib:jar:1.1.3:runtime
[INFO] |  |  |  |  +- com.headius:invokebinder:jar:1.2:runtime
[INFO] |  |  |  |  +- com.martiansoftware:nailgun-server:jar:0.9.1:runtime
[INFO] |  |  |  |  +- org.jruby:yecht:jar:1.0:runtime
[INFO] |  |  |  |  \- com.headius:options:jar:1.1:runtime
[INFO] |  |  |  \- org.jruby:jruby-stdlib:jar:1.7.12:runtime
[INFO] |  |  +- org.scijava:scripting-jython:jar:0.4.1:runtime
[INFO] |  |  |  \- org.scijava:jython-shaded:jar:2.7.0:runtime
[INFO] |  |  +- org.scijava:scripting-renjin:jar:0.2.2:runtime
[INFO] |  |  |  \- org.renjin:renjin-script-engine:jar:0.8.1906:runtime
[INFO] |  |  |     +- org.renjin:renjin-core:jar:0.8.1906:runtime
[INFO] |  |  |     |  +- org.apache.commons:commons-math:jar:2.2:runtime
[INFO] |  |  |     |  +- org.renjin:gcc-runtime:jar:0.8.1906:runtime
[INFO] |  |  |     |  +- org.netlib:netlib-java:jar:0.9.3-renjin-patched-2:runtime
[INFO] |  |  |     |  |  +- org.netlib:lapack:jar:0.8:runtime
[INFO] |  |  |     |  |  +- org.netlib:f2jutil:jar:0.8:runtime
[INFO] |  |  |     |  |  +- org.netlib:xerbla:jar:0.8:runtime
[INFO] |  |  |     |  |  \- org.netlib:blas:jar:0.8:runtime
[INFO] |  |  |     |  +- org.apache.commons:commons-vfs2:jar:2.0:runtime
[INFO] |  |  |     |  |  +- org.apache.maven.scm:maven-scm-api:jar:1.4:runtime
[INFO] |  |  |     |  |  |  \- org.codehaus.plexus:plexus-utils:jar:1.5.6:runtime
[INFO] |  |  |     |  |  \- org.apache.maven.scm:maven-scm-provider-svnexe:jar:1.4:runtime
[INFO] |  |  |     |  |     +- org.apache.maven.scm:maven-scm-provider-svn-commons:jar:1.4:runtime
[INFO] |  |  |     |  |     \- regexp:regexp:jar:1.3:runtime
[INFO] |  |  |     |  +- org.apache.commons:commons-compress:jar:1.4.1:runtime
[INFO] |  |  |     |  |  \- org.tukaani:xz:jar:1.0:runtime
[INFO] |  |  |     |  +- edu.emory.mathcs:jtransforms:jar:2.4:runtime
[INFO] |  |  |     |  +- net.sf.jung:jung-api:jar:2.0.1:runtime
[INFO] |  |  |     |  |  \- net.sourceforge.collections:collections-generic:jar:4.01:runtime
[INFO] |  |  |     |  +- net.sf.jung:jung-graph-impl:jar:2.0.1:runtime
[INFO] |  |  |     |  \- com.sun.codemodel:codemodel:jar:2.6:runtime
[INFO] |  |  |     +- org.renjin:stats:jar:0.8.1906:runtime
[INFO] |  |  |     |  +- org.renjin:renjin-appl:jar:0.8.1906:runtime
[INFO] |  |  |     |  \- org.renjin:renjin-gnur-runtime:jar:0.8.1906:runtime
[INFO] |  |  |     +- org.renjin:methods:jar:0.8.1906:runtime
[INFO] |  |  |     |  \- org.renjin:datasets:jar:0.8.1906:runtime
[INFO] |  |  |     +- org.renjin:utils:jar:0.8.1906:runtime
[INFO] |  |  |     +- org.renjin:grDevices:jar:0.8.1906:runtime
[INFO] |  |  |     \- org.renjin:graphics:jar:0.8.1906:runtime
[INFO] |  |  \- org.scijava:scripting-scala:jar:0.2.1:runtime
[INFO] |  |     \- org.scala-lang:scala-compiler:jar:2.12.1:runtime
[INFO] |  |        +- org.scala-lang:scala-library:jar:2.12.1:runtime
[INFO] |  |        +- org.scala-lang:scala-reflect:jar:2.12.1:runtime
[INFO] |  |        \- org.scala-lang.modules:scala-xml_2.12:jar:1.0.6:runtime
[INFO] |  +- net.imagej:imagej-ui-swing:jar:0.21.2:runtime
[INFO] |  |  +- com.miglayout:miglayout:jar:swing:3.7.4:runtime
[INFO] |  |  +- org.jhotdraw:jhotdraw:jar:7.6.0:runtime
[INFO] |  |  \- net.imagej:imagej-ui-awt:jar:0.3.1:runtime
[INFO] |  +- org.scijava:scijava-plugins-commands:jar:0.2.3:runtime
[INFO] |  +- org.scijava:scijava-plugins-platforms:jar:0.3.0:runtime
[INFO] |  +- org.scijava:scijava-plugins-text-markdown:jar:0.1.3:runtime
[INFO] |  |  \- org.markdownj:markdownj:jar:0.3.0-1.0.2b4:runtime
[INFO] |  +- org.scijava:scijava-plugins-text-plain:jar:0.1.3:runtime
[INFO] |  \- org.scijava:script-editor:jar:0.1.4:runtime
[INFO] |     +- com.fifesoft:rsyntaxtextarea:jar:2.6.1:runtime
[INFO] |     \- com.fifesoft:languagesupport:jar:2.6.0:runtime
[INFO] |        +- com.fifesoft:autocomplete:jar:2.6.1:runtime
[INFO] |        \- org.mozilla:rhino:jar:1.7.6:runtime
[INFO] +- net.imagej:ij:jar:1.51n:compile
[INFO] +- sc.fiji:TrackMate_:jar:3.5.3:compile
[INFO] |  +- sc.fiji:3D_Viewer:jar:4.0.2:compile
[INFO] |  |  \- sc.fiji:VIB-lib:jar:2.1.2:compile
[INFO] |  |     \- sc.fiji:pal-optimization:jar:2.0.1:compile
[INFO] |  +- sc.fiji:fiji-lib:jar:2.1.2:compile
[INFO] |  +- net.imglib2:imglib2:jar:4.2.1:compile
[INFO] |  +- net.imglib2:imglib2-algorithm:jar:0.8.1:compile
[INFO] |  |  +- net.sf.trove4j:trove4j:jar:3.0.3:compile
[INFO] |  |  \- org.ojalgo:ojalgo:jar:43.0:compile
[INFO] |  +- net.imglib2:imglib2-algorithm-gpl:jar:0.2.1:compile
[INFO] |  |  +- mpicbg:mpicbg:jar:1.1.1:compile
[INFO] |  |  \- edu.mines:mines-jtk:jar:20151125:compile
[INFO] |  +- net.imglib2:imglib2-ij:jar:2.0.0-beta-38:compile
[INFO] |  +- net.imglib2:imglib2-realtransform:jar:2.0.0-beta-38:compile
[INFO] |  |  \- jitk:jitk-tps:jar:3.0.0:compile
[INFO] |  |     +- com.googlecode.efficient-java-matrix-library:ejml:jar:0.24:compile
[INFO] |  |     \- log4j:log4j:jar:1.2.17:compile
[INFO] |  +- org.scijava:j3dcore:jar:1.6.0-scijava-2:compile
[INFO] |  |  +- org.jogamp.gluegen:gluegen-rt:jar:2.3.2:compile
[INFO] |  |  \- org.jogamp.jogl:jogl-all:jar:2.3.2:compile
[INFO] |  +- org.scijava:j3dutils:jar:1.6.0-scijava-2:compile
[INFO] |  +- org.scijava:vecmath:jar:1.6.0-scijava-2:compile
[INFO] |  +- com.itextpdf:itextpdf:jar:5.5.11:compile
[INFO] |  +- gov.nist.math:jama:jar:1.0.3:compile
[INFO] |  +- net.sf.jgrapht:jgrapht:jar:0.8.3:compile
[INFO] |  +- org.apache.xmlgraphics:batik-dom:jar:1.9:compile
[INFO] |  |  +- org.apache.xmlgraphics:batik-css:jar:1.9:compile
[INFO] |  |  |  \- org.apache.xmlgraphics:xmlgraphics-commons:jar:2.2:compile
[INFO] |  |  +- org.apache.xmlgraphics:batik-ext:jar:1.9:compile
[INFO] |  |  +- org.apache.xmlgraphics:batik-util:jar:1.9:compile
[INFO] |  |  +- org.apache.xmlgraphics:batik-xml:jar:1.9:compile
[INFO] |  |  \- xml-apis:xml-apis-ext:jar:1.3.04:compile
[INFO] |  +- org.apache.xmlgraphics:batik-svggen:jar:1.9:compile
[INFO] |  |  \- org.apache.xmlgraphics:batik-awt-util:jar:1.9:compile
[INFO] |  +- org.apache.xmlgraphics:batik-constants:jar:1.9:compile
[INFO] |  +- org.apache.xmlgraphics:batik-i18n:jar:1.9:compile
[INFO] |  +- org.jdom:jdom2:jar:2.0.6:compile
[INFO] |  +- org.jfree:jcommon:jar:1.0.24:compile
[INFO] |  +- org.jfree:jfreechart:jar:1.0.19:compile
[INFO] |  \- org.tinyjee.jgraphx:jgraphx:jar:3.4.1.3:compile
[INFO] +- org.nd4j:nd4j-api:jar:0.9.1:compile
[INFO] |  +- org.slf4j:slf4j-api:jar:1.7.25:compile
[INFO] |  +- org.nd4j:jackson:jar:0.9.1:compile
[INFO] |  |  +- org.yaml:snakeyaml:jar:1.12:compile
[INFO] |  |  +- org.codehaus.woodstox:stax2-api:jar:3.1.4:compile
[INFO] |  |  \- joda-time:joda-time:jar:2.9.9:compile
[INFO] |  +- org.projectlombok:lombok:jar:1.16.16:compile
[INFO] |  +- org.nd4j:nd4j-buffer:jar:0.9.1:compile
[INFO] |  +- org.nd4j:nd4j-context:jar:0.9.1:compile
[INFO] |  +- net.ericaro:neoitertools:jar:1.0.0:compile
[INFO] |  \- org.reflections:reflections:jar:0.9.10:compile
[INFO] |     \- com.google.code.findbugs:annotations:jar:2.0.1:compile
[INFO] +- org.nd4j:nd4j-common:jar:0.9.1:compile
[INFO] |  +- commons-io:commons-io:jar:2.4:compile
[INFO] |  +- org.apache.commons:commons-math3:jar:3.6.1:compile
[INFO] |  \- org.apache.commons:commons-lang3:jar:3.5:compile
[INFO] +- org.nd4j:nd4j-native:jar:0.9.1:compile
[INFO] |  +- org.nd4j:nd4j-native:jar:linux-x86_64:0.9.1:compile
[INFO] |  +- org.bytedeco:javacpp:jar:1.3.3:compile
[INFO] |  +- org.bytedeco.javacpp-presets:openblas:jar:0.2.19-1.3:compile
[INFO] |  +- org.bytedeco.javacpp-presets:openblas:jar:linux-x86_64:0.2.19-1.3:compile
[INFO] |  \- org.nd4j:nd4j-native-api:jar:0.9.1:compile
[INFO] +- org.nd4j:nd4j-native-platform:jar:0.9.1:compile
[INFO] |  +- org.bytedeco.javacpp-presets:openblas-platform:jar:0.2.19-1.3:compile
[INFO] |  |  +- org.bytedeco.javacpp-presets:openblas:jar:android-arm:0.2.19-1.3:compile
[INFO] |  |  +- org.bytedeco.javacpp-presets:openblas:jar:android-x86:0.2.19-1.3:compile
[INFO] |  |  +- org.bytedeco.javacpp-presets:openblas:jar:linux-x86:0.2.19-1.3:compile
[INFO] |  |  +- org.bytedeco.javacpp-presets:openblas:jar:linux-armhf:0.2.19-1.3:compile
[INFO] |  |  +- org.bytedeco.javacpp-presets:openblas:jar:linux-ppc64le:0.2.19-1.3:compile
[INFO] |  |  +- org.bytedeco.javacpp-presets:openblas:jar:macosx-x86_64:0.2.19-1.3:compile
[INFO] |  |  +- org.bytedeco.javacpp-presets:openblas:jar:windows-x86:0.2.19-1.3:compile
[INFO] |  |  \- org.bytedeco.javacpp-presets:openblas:jar:windows-x86_64:0.2.19-1.3:compile
[INFO] |  +- org.nd4j:nd4j-native:jar:android-arm:0.9.1:compile
[INFO] |  +- org.nd4j:nd4j-native:jar:android-x86:0.9.1:compile
[INFO] |  +- org.nd4j:nd4j-native:jar:macosx-x86_64:0.9.1:compile
[INFO] |  +- org.nd4j:nd4j-native:jar:windows-x86_64:0.9.1:compile
[INFO] |  \- org.nd4j:nd4j-native:jar:linux-ppc64le:0.9.1:compile
[INFO] +- org.slf4j:slf4j-nop:jar:1.7.25:compile
[INFO] +- de.biomedical-imaging.imagej:ij_ridge_detect:jar:1.4.1:compile
[INFO] +- com.google.code.gson:gson:jar:2.8.2:compile
[INFO] +- com.opencsv:opencsv:jar:4.0:compile
[INFO] |  +- org.apache.commons:commons-text:jar:1.1:compile
[INFO] |  \- commons-beanutils:commons-beanutils:jar:1.9.3:compile
[INFO] |     +- commons-logging:commons-logging:jar:1.2:compile
[INFO] |     \- commons-collections:commons-collections:jar:3.2.2:compile
[INFO] +- sc.fiji:KymographBuilder:jar:2.0.0:compile
[INFO] |  \- net.imagej:ij1-patcher:jar:0.12.6:compile
[INFO] +- com.google.guava:guava:jar:21.0:compile
[INFO] +- net.imagej:imagej-legacy:jar:0.25.0:runtime (optional) 
[INFO] |  +- net.imglib2:imglib2-roi:jar:0.4.6:compile
[INFO] |  +- org.scijava:scijava-ui-awt:jar:0.1.6:runtime
[INFO] |  +- org.scijava:scijava-ui-swing:jar:0.9.3:runtime
[INFO] |  |  +- org.scijava:swing-checkbox-tree:jar:1.0.2:runtime
[INFO] |  |  \- net.sourceforge.jdatepicker:jdatepicker:jar:1.3.2:runtime
[INFO] |  \- org.javassist:javassist:jar:3.22.0-CR1:compile
[INFO] \- junit:junit:jar:4.12:test
[INFO]    \- org.hamcrest:hamcrest-core:jar:1.3:test
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 2.066 s
[INFO] Finished at: 2017-11-07T17:53:51-05:00
[INFO] Final Memory: 31M/457M
[INFO] ------------------------------------------------------------------------
```

**Comment by ctrueden** [comment]
@hadim Maybe the problem is that you [start your own thread](https://github.com/hadim/FilamentDetector/blob/60723ae94dd9da0a0e06bf34aa25872cd1de3b6f/src/main/java/fiji/plugin/filamentdetector/gui/controller/AnalyzeController.java#L257), but without setting the thread's context class loader. So the IJ1 class loader is not used. This may be a problem.

Try adding the following line at that point in your code:
```java
analyzerThread.setContextClassLoader(IJ.getClassLoader());
```

Or to avoid a reference to `ij.IJ`, you could try:
```java
analyzerThread.setContextClassLoader(Thread.currentThread().getContextClassLoader());
```

I also may be totally wrong about all of this, but it is suspicious that the ND4J code is trying to call `ServiceLoader.load` which uses the context class loader by default.

You can also double check that the JAR file(s) containing the desired backends do indeed have `META-INF/services` folders containing the metadata about those services. But I'd be surprised if they don't. After all, it works from Eclipse.

Another thing you can try is to launch ImageJ from the CLI with `--full-classpath` option, which will put all JARs in jars and plugins on the system classpath, which I bet you will avoid this issue.

**Comment by agibsonccc** [comment]
@ctrueden we do this to dynamically load native binaries at runtime. Here is the code we use to do that: https://github.com/deeplearning4j/nd4j/blob/master/nd4j-context/src/main/java/org/nd4j/linalg/factory/Nd4jBackend.java

**Comment by ctrueden** [comment]
@agibsonccc Indeed! That is the code I was looking at (specifically, [this line](https://github.com/deeplearning4j/nd4j/blob/0bc4e2224b1b9e4405b6f3111761956d4224439e/nd4j-context/src/main/java/org/nd4j/linalg/factory/Nd4jBackend.java#L154)) which led me to check the `java.util.ServiceLoader` code, which I noticed used the current thread's context class loader, which reminded me that ImageJ 1.x is weird about that, which led me to check @hadim's code to see if/how he is using threads, and indeed I saw that he creates a new thread and launches it.

@hadim Please keep us posted on what you discover—happy to dig further if you get stuck.

**Comment by saudet** [comment]
Could you try setting the "org.bytedeco.javacpp.logger.debug" to "true" before and post the additional output that gets generated?

**Comment by ctrueden** [comment]
I am debugging into this now on the ImageJ side. Will report back again with findings shortly.

MCVE here: https://github.com/ctrueden/nd4j-plugin-mcve

Here is the output when the main method is run from Eclipse:
```
Creating ND4J thing
SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
Loading class org.nd4j.nativeblas.Nd4jCpu
Loading class org.nd4j.nativeblas.Nd4jCpu
Loading library libnd4jcpu
Failed to load for libnd4jcpu: java.lang.UnsatisfiedLinkError: no libnd4jcpu in java.library.path
Loading /Users/curtis/.javacpp/cache/nd4j-native-0.9.1-macosx-x86_64.jar/org/nd4j/nativeblas/macosx-x86_64/libnd4jcpu.dylib
Loading /Users/curtis/.javacpp/cache/nd4j-native-0.9.1-macosx-x86_64.jar/org/nd4j/nativeblas/macosx-x86_64/libjnind4jcpu.dylib
Loading library libnd4jcpu
Failed to load for libnd4jcpu: java.lang.UnsatisfiedLinkError: no libnd4jcpu in java.library.path
Registering org.nd4j.nativeblas.Nd4jCpu$NativeOps[address=0x7ff1bb018ea0,position=0,limit=1,capacity=1,deallocator=org.bytedeco.javacpp.Pointer$NativeDeallocator[ownerAddress=0x7ff1bb018ea0,deallocatorAddress=0x180871940]]
Loading class org.bytedeco.javacpp.openblas
Loading library iomp5
Failed to load for iomp5: java.lang.UnsatisfiedLinkError: no iomp5 in java.library.path
Loading library mkl_avx
Failed to load for mkl_avx: java.lang.UnsatisfiedLinkError: no mkl_avx in java.library.path
Loading library mkl_avx2
Failed to load for mkl_avx2: java.lang.UnsatisfiedLinkError: no mkl_avx2 in java.library.path
Loading library mkl_avx512_mic
Failed to load for mkl_avx512_mic: java.lang.UnsatisfiedLinkError: no mkl_avx512_mic in java.library.path
Loading library mkl_def
Failed to load for mkl_def: java.lang.UnsatisfiedLinkError: no mkl_def in java.library.path
Loading library mkl_mc3
Failed to load for mkl_mc3: java.lang.UnsatisfiedLinkError: no mkl_mc3 in java.library.path
Loading library mkl_core
Failed to load for mkl_core: java.lang.UnsatisfiedLinkError: no mkl_core in java.library.path
Loading library mkl_gnu_thread
Failed to load for mkl_gnu_thread: java.lang.UnsatisfiedLinkError: no mkl_gnu_thread in java.library.path
Loading library mkl_intel_lp64
Failed to load for mkl_intel_lp64: java.lang.UnsatisfiedLinkError: no mkl_intel_lp64 in java.library.path
Loading library mkl_intel_thread
Failed to load for mkl_intel_thread: java.lang.UnsatisfiedLinkError: no mkl_intel_thread in java.library.path
Loading library mkl_rt
Failed to load for mkl_rt: java.lang.UnsatisfiedLinkError: no mkl_rt in java.library.path
Loading library mkl_rt
Failed to load for mkl_rt#openblas@.0: java.lang.UnsatisfiedLinkError: no mkl_rt in java.library.path
Loading library gcc_s
Failed to load for gcc_s@.1: java.lang.UnsatisfiedLinkError: no gcc_s in java.library.path
Loading library quadmath
Failed to load for quadmath@.0: java.lang.UnsatisfiedLinkError: no quadmath in java.library.path
Loading library gfortran
Failed to load for gfortran@.3: java.lang.UnsatisfiedLinkError: no gfortran in java.library.path
Loading /Users/curtis/.javacpp/cache/openblas-0.2.19-1.3-macosx-x86_64.jar/org/bytedeco/javacpp/macosx-x86_64/libopenblas.dylib
Loading /Users/curtis/.javacpp/cache/openblas-0.2.19-1.3-macosx-x86_64.jar/org/bytedeco/javacpp/macosx-x86_64/libjniopenblas.dylib
Registering org.bytedeco.javacpp.FloatPointer[address=0x7ff1c062abe0,position=0,limit=100,capacity=100,deallocator=org.bytedeco.javacpp.Pointer$NativeDeallocator[ownerAddress=0x7ff1c062abe0,deallocatorAddress=0x180871a70]]
Registering org.bytedeco.javacpp.IntPointer[address=0x7ff1b8f1fbd0,position=0,limit=8,capacity=8,deallocator=org.bytedeco.javacpp.Pointer$NativeDeallocator[ownerAddress=0x7ff1b8f1fbd0,deallocatorAddress=0x180871a30]]
Created ND4J thing
[WARNING] Ignoring unsupported output: result [org.nd4j.linalg.cpu.nativecpu.NDArray]
```

Here is the output when the code runs from ImageJ:
```
Creating ND4J thing
Reflections took 129 ms to scan 10 urls, producing 373 keys and 1736 values 
[ERROR] Module threw error
java.lang.ExceptionInInitializerError
	at nd4j_plugin.ND4JPlugin.run(ND4JPlugin.java:22)
	at org.scijava.command.CommandModule.run(CommandModule.java:199)
	at org.scijava.module.ModuleRunner.run(ModuleRunner.java:168)
	at org.scijava.module.ModuleRunner.call(ModuleRunner.java:127)
	at org.scijava.module.ModuleRunner.call(ModuleRunner.java:66)
	at org.scijava.thread.DefaultThreadService$3.call(DefaultThreadService.java:238)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.RuntimeException: org.nd4j.linalg.factory.Nd4jBackend$NoAvailableBackendException: Please ensure that you have an nd4j backend on your classpath. Please see: http://nd4j.org/getstarted.html
	at org.nd4j.linalg.factory.Nd4j.initContext(Nd4j.java:6089)
	at org.nd4j.linalg.factory.Nd4j.<clinit>(Nd4j.java:201)
	... 10 more
Caused by: org.nd4j.linalg.factory.Nd4jBackend$NoAvailableBackendException: Please ensure that you have an nd4j backend on your classpath. Please see: http://nd4j.org/getstarted.html
	at org.nd4j.linalg.factory.Nd4jBackend.load(Nd4jBackend.java:258)
	at org.nd4j.linalg.factory.Nd4j.initContext(Nd4j.java:6086)
	... 11 more
```

**Comment by saudet** [comment]
Looks like ImageJ is loading an old version of JavaCPP somewhere...

**Comment by ctrueden** [comment]
TL;DR This is not a bug in ND4J. So this issue can be closed.

I was wrong: the class loader was not the culprit. The same problem still occurs with ImageJ's class loader and the system class loader.

It was actually a much simpler issue: there were two JAR files missing from the ImageJ installation: `nd4j-native-0.9.1.jar` and `openblas-0.2.19-1.3.jar`.

This is highly likely to be a bug in the `imagej-maven-plugin` which is responsible for copying dependencies from a project into an ImageJ installation. Notice that both of those artifacts also have other related artifacts: multiple native classifier suffixes. I think the `imagej-maven-plugin` is getting confused about this.

In case you are curious: I figured this out by using ImageJ's "System Information" command, which dumps a bunch of info, including the Java classpath. I saved the output of the "bad" ImageJ to a file. I repeated the process for the "good" ImageJ launched from Eclipse. Then I stripped out all the stuff except the respective Java classpaths, and used the following for loop:
```bash
for f in $(cat bad-ij-jars.txt)
do
services=$(jar tf "$f" | grep 'META-INF/services.*Nd4jBackend')
test "$services" && echo "[$f] $services"
done
```
and same for `good-ij-jars.txt`.

For bad ImageJ, I got no output. For the good ImageJ from Eclipse, I got a hit:
```
[/Users/curtis/.m2/repository/org/nd4j/nd4j-native/0.9.1/nd4j-native-0.9.1.jar] META-INF/services/org.nd4j.linalg.factory.Nd4jBackend
```
And then I checked for `nd4j-native-0.9.1.jar` in the real/bad ImageJ installation and found it missing. With continued investigation I found `openblas-0.2.19-1.3.jar` also missing. Adding both of those JAR files caused my MCVE to start working.
